compiler/rustc_lint/src/types/improper_ctypes.rs

```diff
@@ -49,3 +97,1385 @@ fn variant_has_complex_ctor(variant: &ty::VariantDef) -> bool {
+                // (which has a stable layout) but points to FFI-unsafe type, is it safe?
+                // On one hand, the function's ABI will match that of a similar C-declared function API,
+                // on the other, dereferencing the pointer on the other side of the FFI boundary will be painful.
+                // In this code, the opinion on is split between function declarations and function definitions,
+                // with the idea that at least one side of the FFI boundary needs to treat the pointee as an opaque type.
+                // For declarations, we see this as unsafe, but for definitions, we see this as safe.
+                //
+                // For extern function declarations, the actual definition of the function is written somewhere else,
+                // meaning the declaration is free to express this opaqueness with an extern type (opaque caller-side) or a std::ffi::c_void (opaque callee-side)
+                // (or other possibly better tricks, see https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs)
+                // For extern function definitions, however, in the case where the type is opaque caller-side, it is not opaque callee-side,
+                // and having the full type information is necessary to compile the function.
+                if state.is_in_defined_function() {
+                    FfiResult::FfiSafe
+                } else {
+                    return self.visit_type(state, Some(ty), inner_ty).forbid_phantom().wrap_all(
+                        ty,
+                        fluent::lint_improper_ctypes_sized_ptr_to_unsafe_type,
+                        None,
+                    );
+                }
+            }
+            TypeSizedness::NotYetKnown => {
+                // types with sizedness NotYetKnown:
+                // - Type params (with `variable: impl Trait` shorthand or not)
+                //   (function definitions only, let's see how this interacts with monomorphisation)
+                // - Self in trait functions/methods
+                //   (FIXME note: function 'declarations' there should be treated as definitions)
+                // - Opaque return types
+                //   (always FFI-unsafe)
+                // - non-exhaustive structs/enums/unions from other crates
+                //   (always FFI-unsafe)
+                // (for the three first, this is unless there is a `+Sized` bound involved)
+                //
+                // FIXME: on a side note, we should separate 'true' declarations (non-rust code),
+                // 'fake' declarations (in traits, needed to be implemented elsewhere), and definitions.
+                // (for instance, definitions should worry about &self with Self:?Sized, but fake declarations shouldn't)
+
+                // wether they are FFI-safe or not does not depend on the indirections involved (&Self, &T, Box<impl Trait>),
```

Comment:
Simpler spellcheckers cannot catch this because "wether" is in fact a word (though definitely not the one you meant), and more complex ones are infuriating because they are never quite smart enough to understand when you meant to use a word that way.
```suggestion
                // whether they are FFI-safe or not does not depend on the indirections involved (&Self, &T, Box<impl Trait>),
```